misk/src/main/kotlin/misk/hibernate/ReflectionQueryFactory.kt

```diff
@@ -0,0 +1,257 @@
+package misk.hibernate
+
+import com.google.common.cache.CacheBuilder
+import com.google.common.cache.CacheLoader
+import misk.inject.typeLiteral
+import java.lang.reflect.InvocationHandler
+import java.lang.reflect.Method
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Proxy
+import javax.inject.Singleton
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Path
+import javax.persistence.criteria.Predicate
+import javax.persistence.criteria.Root
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.full.allSupertypes
+import kotlin.reflect.full.declaredMemberFunctions
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.javaMethod
+
+/**
+ * Implements the [Query] @[Constraint] methods using a dynamic proxy, and projections using
+ * reflection to call the primary constructor.
+ */
+internal class ReflectionQuery<T : DbEntity<T>>(
+  private val factory: Factory,
+  private val rootEntityType: KClass<T>,
+  private val queryMethodHandlers: Map<Method, QueryMethodHandler>
+) : Query<T>, InvocationHandler {
+  private val constraints = mutableListOf<ConstraintSpec>()
+
+  override fun uniqueResult(session: Session): T? {
+    // TODO(jwilson): max results = 2
```

Comment:
this is to put a limit of 2 in the query for safety?

Reply:
yeah, so if the DB returns 1 result you’re good, and if it returns more than one you know your invariant was off